train/train_process.py

```diff
@@ -173,103 +176,144 @@ def _self_play(self, model_path):
-            elif losing > self.cfg["model_compare"]["warning_losing_threshold"]:
-                logging.warn("New model is worse than previous one, losing ratio: %f", losing)
-
-            return best_model
-        finally:
-            shutil.rmtree(tmp_dirpath)
+            return latest_models[0]
+
+        logging.info("Comparing latest models to best model...")
+        for (model_idx, (latest_model, latest_model_path)) in enumerate(latest_models):
+            tmp_dirpath = tempfile.mkdtemp()
+            try:
+                compare_res_file = os.path.join(tmp_dirpath, "compare_result.json")
+                tmp_games_dir = os.path.join(tmp_dirpath, "games")
+
+                profile = "dev" if self.cfg["debug"] else "release"
+                games_dir = os.path.join(self.cfg["games_dir"], self.run_id)
+                data_entries_dir = os.path.join(games_dir, datetime.datetime.now().strftime("%y%m%d_%H%M%S"))
+
+                compare_start_time = time.time()
+                subprocess.run(prepare_cmd(
+                    "cargo", "run", "--profile", profile, "-q", "--bin",
+                    self.self_play_exec, "--",
+                    "--model1-path", best_model[1],
+                    "--model2-path", latest_model_path,
+                    "--games-num", self.cfg["model_compare"]["games_num"],
+                    # take the opportunity to generate more games to main games directory
+                    "--out-dir1", data_entries_dir,
+                    "--out-dir2", tmp_games_dir,
+                    "--summary-file", compare_res_file,
+                    "--sim-num", self.cfg["mcts"]["sim_num"],
+                    "--explore-factor", self.cfg["mcts"]["explore_factor"],
+                    "--temperature-policy", self.cfg["model_compare"]["temperature_policy_str"],
+                    "--prior-noise-alpha", self.cfg["mcts"]["prior_noise_alpha"],
+                    "--prior-noise-epsilon", self.cfg["mcts"]["prior_noise_epsilon"],
+                    "--threads", self.cfg["model_compare"]["threads"],
+                    "--processing-unit", "CPU" if self.cfg["cpu"] else "GPU"),
+                    stderr=sys.stderr, stdout=sys.stdout, shell=True, check=True)
+                self.metrics["model_compare_duration"] = time.time() - compare_start_time
+
+                with open(compare_res_file, "r") as res_file:
+                    res = json.load(res_file)
+                w1, w2, d = res["player1_wins"], res["player2_wins"], res["draws"]
+                total_games = w1 + w2 + d
+                winning, losing = w2 / total_games, w1 / total_games
+                self.metrics[f"trained_model_win_rate_{model_idx}"] = winning
+                logging.debug(f"Trained model winning rate: {winning}")
+
+                if winning > self.cfg["model_compare"]["switching_winning_threshold"]:
+                    best_model = (latest_model, latest_model_path)

```

Comment:
this is kind of a like a simple tournament? 
where you always play against the current best, and if you lose once, you're out? 
i thought the plan was letting all the models play against the original, and then see who won most.

The advantage of this latter approach is that it does not depend on the order of the games. Also it makes the metrics more intuitive because all the models in a single iteration are compared against the same base.
But I don't know if that matters so much - what do you think will be best for maximizing the model quality?

Reply:
not exactly a tournament, simply:

winning=compare(model, best)
if winning > 0.55: best = model

Every trained model is 'out' after playing.
You are correct that 'winning_rate' is less intuitive because we may switch the best model during comparison, but we always want to compare to the real best, it will be weird to compare to an older 'best'.
I think that the best way to find the best one